Cython/Compiler/ExprNodes.py

```diff
@@ -520,6 +536,11 @@ def result_as(self, type = None):
             return typecast(type, py_object_type, self.result())
         return typecast(type, self.ctype(), self.result())
 
+    def move_result_as_rhs(self, type = None):
+        allow_move = True if type is None else (not type.is_reference)
```

Comment:
```suggestion
        allow_move = type is None or not type.is_reference
```
The interface feels a bit unsafe here. Why would we not have a type? Why can we safely `move` if no type is provided (or known)?

Reply:
It does seem the wrong way round. I've changed it to assuming we can't move if type is `None`/unknown. I've also excluded types with `needs_refcounting` (which didn't exist when this PR was created, but I think makes sense).

I think this PR is more about adding a mechanism to support moving temps rather than covering every possible case, so it's better to be cautious now and use it more later if it turns out to make sense.